File: pytsn/vlan.py
import shlex
import subprocess
import sys


def run_cmd(cmd: str):
    print(cmd, file=sys.stderr)
    return subprocess.check_call(shlex.split(cmd))


def vlan_name(ifname: str, vlanid) -> str:
    return f'{ifname}.{vlanid}'


def setup_tas(ifname: str, tas: dict):
    handle = 100
    num_tc = tas['num_tc']
    priomap = ' '.join(map(str, tas['tc_map']))
    queues = ' '.join(tas['queues'])
    base_time = tas['base_time']
    sched_entries = ' '.join(f'sched-entry {entry}' for entry in tas['sched_entries'])
    txtime_delay = tas['txtime_delay']

    run_cmd(
        f'tc qdisc replace dev {ifname} parent root handle {handle} taprio '
        f'num_tc {num_tc} '
        f'map {priomap} '
        f'queues {queues} '
        f'base-time {base_time} '
        f'{sched_entries} '
        f'flags 0x0 '
        f'txtime-delay {txtime_delay} '
        f'clockid CLOCK_TAI'
    )

    run_cmd(
        f'tc qdisc replace dev {ifname} parent {handle}:1 etf '
        'clockid CLOCK_TAI '
        f'delta {txtime_delay} '
        'skip_sock_check'
    )


def setup_cbs(ifname: str, cbs: dict):
    root_handle = 100
    num_tc = cbs['num_tc']
    priomap = ' '.join(map(str, cbs['tc_map']))
    queues = ' '.join(cbs['queues'])
    run_cmd(
        f'tc qdisc add dev {ifname} parent root '
        f'handle {root_handle} mqprio '
        f'num_tc {num_tc} '
        f'map {priomap} '
        f'queues {queues} '
        f'hw 0'
    )

    for qid, val in cbs['children'].items():
        handle = qid * 1111

        idleslope = val['idleslope']
        sendslope = val['sendslope']
        hicredit = val['hicredit']
        locredit = val['locredit']
        run_cmd(
            f'tc qdisc replace dev {ifname} '
            f'parent {root_handle}:{qid} '
            f'handle {handle} cbs '
            f'idleslope {idleslope} '
            f'sendslope {sendslope} '
            f'hicredit {hicredit} '
            f'locredit {locredit} '
            f'offload 0'
        )
        # run_cmd(
        #     f'tc qdisc add dev {ifname} parent {handle}:1 etf '
        #     'clockid CLOCK_TAI '
        #     'delta 500000 '
        #     'offload '
        # )


def create_vlan(config: dict, ifname: str, vlanid: int) -> int:
    name = vlan_name(ifname, vlanid)

    try:
        ifconf = config['nics'][ifname]

        # Not support tas+cbs yet
        if all(x in ifconf for x in ('tas', 'cbs')):
            raise ValueError('Does not support tas + cbs yet')

        qos_map = ' '.join(
            f'{skb_pri}:{vlan_pri}'
            for skb_pri, vlan_pri
            in ifconf['vlan'][vlanid]['maps'].items())

        ipv4 = ifconf['vlan'][vlanid].get('ipv4', None)

        run_cmd(
            f'ip link add link {ifname} name {name} type vlan id {vlanid} '
            f'egress-qos-map {qos_map}'
        )

        run_cmd(
            f'ip link set up {name}'
        )

        if ipv4:
            run_cmd(
                f'ip addr add {ipv4} dev {name}'
            )

        if 'tas' in ifconf:
            setup_tas(ifname, ifconf['tas'])

        if 'cbs' in ifconf:
            setup_cbs(ifname, ifconf['cbs'])
    except subprocess.CalledProcessError as e:
        return e.returncode
    except KeyError as e:
        print(e)
        print('Config is not properly configuered', file=sys.stderr)
        return 1
    else:
        return 0


def delete_vlan(config: dict, ifname: str, vlanid: int) -> int:
    name = vlan_name(ifname, vlanid)
    try:
        run_cmd(
            f'ip link del {name}'
        )

        run_cmd(
            f'tc qdisc delete dev {ifname} root'
        )
    except subprocess.CalledProcessError as e:
        return e.returncode
    else:
        return 0

File: src/vlan.rs
use crate::{cbs::CbsConfig, config::Config, tas::TasConfig};
use itertools::Itertools;
use std::{collections::HashMap, io::Error};

fn run_cmd(input: &str) -> Result<i32, String> {
    eprintln!("{}", input);
    let split = input.split(char::is_whitespace);
    let cmd = split.clone().next().unwrap();
    let cmd = std::process::Command::new(cmd)
        .args(split.skip(1))
        .spawn()
        .unwrap();
    let output = cmd.wait_with_output().unwrap();
    if output.status.success() {
        Ok(0)
    } else {
        Err(Error::from_raw_os_error(output.status.code().unwrap()).to_string())
    }
}

pub fn setup_tas(ifname: &str, config: &TasConfig) -> Result<i32, String> {
    let handle = 100;
    let num_tc = config.num_tc;
    let mut priomap = String::new();
    let mut queues = String::new();
    let base_time = config.base_time;
    let txtime_delay = config.txtime_delay;
    let mut sched_entries = String::new();
    for key in config.tc_map.keys().sorted() {
        priomap.push_str(&format!(" {}", config.tc_map.get(key).unwrap()));
    }
    for queue in &config.queues {
        queues.push_str(&format!(" {}", queue));
    }
    for entry in &config.sched_entries {
        sched_entries.push_str(&format!(" sched-entry {}", entry));
    }
    let cmd = format!(
        "tc qdisc replace dev {} parent root handle {} taprio num_tc {} map{} \
         queues{} base-time {}{} flags 0x2 txtime-delay {}",
        ifname, handle, num_tc, priomap, queues, base_time, sched_entries, txtime_delay
    );
    run_cmd(&cmd)?;
    // TSN NIC does not support ETF for now
    /*
    let cmd = format!(
        "tc qdisc replace dev {} parent {}:1 etf clockid \
         CLOCK_TAI delta {} offload skip_sock_check",
        ifname, handle, txtime_delay
    );
    run_cmd(&cmd)?;
    */
    Ok(0)
}

pub fn setup_cbs(ifname: &str, config: &CbsConfig) -> Result<i32, String> {
    let root_handle = 100;
    let num_tc = config.num_tc;
    let mut priomap = String::new();
    let mut queues = String::new();
    for key in config.tc_map.keys().sorted() {
        priomap.push_str(&format!(" {}", config.tc_map.get(key).unwrap()));
    }
    for s in &config.queues {
        queues.push_str(&format!("{} ", s));
    }
    let cmd = format!(
        "tc qdisc add dev {} parent root handle {} mqprio num_tc {} map \
         {} queues {} hw 0",
        ifname, root_handle, num_tc, priomap, queues
    );
    run_cmd(&cmd)?;
    for (qid, val) in &config.children {
        let handle = qid * 1111;

        let idleslope = val.idleslope;
        let sendslope = val.sendslope;
        let hicredit = val.hicredit;
        let locredit = val.locredit;
        let cmd = format!(
            "tc qdisc replace dev {} parent {}:{} handle {} \
             cbs idleslope {} sendslope {} hicredit {} locredit {} offload 1",
            ifname, root_handle, qid, handle, idleslope, sendslope, hicredit, locredit
        );
        run_cmd(&cmd)?;
    }
    Ok(0)
}

pub fn create_vlan(config: &Config, ifname: &str, vlan_id: u16) -> Result<i32, String> {
    let name = get_vlan_name(ifname, vlan_id);
    let mut qos_map = HashMap::new();

    if config.tas.is_some() && config.cbs.is_some() {
        eprintln!("Does not support both TAS and CBS");
        return Err("Does not support both TAS and CBS".to_string());
    }
    for (prio, pri) in config.egress_qos_map.get(&(vlan_id as i64)).unwrap() {
        qos_map.insert(prio, pri);
    }
    let mut cmd = String::new();
    cmd.push_str(&format!(
        "ip link add link {} name {} type vlan id {} egress-qos-map",
        ifname, name, vlan_id
    ));
    for (prio, pri) in qos_map {
        cmd.push_str(&format!(" {}:{}", pri, prio));
    }
    run_cmd(&cmd)?;
    let cmd = format!("ip link set up {}", name);
    run_cmd(&cmd)?;
    if let Some(tas) = &config.tas {
        setup_tas(ifname, tas)?;
    }
    if let Some(cbs) = &config.cbs {
        setup_cbs(ifname, cbs)?;
    }
    Ok(0)
}

pub fn delete_vlan(ifname: &str, vlanid: u16) -> Result<i32, String> {
    let name = get_vlan_name(ifname, vlanid);
    let cmd = format!("ip link del {}", name);
    run_cmd(&cmd)?;
    let cmd = format!("tc qdisc delete dev {} root", ifname);
    run_cmd(&cmd)?;
    Ok(0)
}

pub fn get_vlan_name(ifname: &str, vlanid: u16) -> String {
    if ifname.len() > 10 {
        format!("{}.{}", &ifname[..10], vlanid)
    } else {
        format!("{}.{}", &ifname, vlanid)
    }
}
